### PR TITLE
[SYCL-MLIR][cgeist] Allow SYCL types not in the `sycl` dialect by default

### PR DESCRIPTION
**Pull request by victor-eds**
This should be an opt-in feature for testing, not something we want to have when using the compiler. Lately usage of SYCL-MLIR to build SYCL projects lead to adding this option just to allow SYCL types not in the dialect yet or those we do not even want to have, e.g., deprecated features.

cgeist SYCL codegen tests use the new `-forbid-undefined-sycl-types` option to test dialect coverage.

### CODE REVIEW

**Review by whitneywhtsang — COMMENTED** [review]
With this change, we may unexpectedly allow types in sycl namespace not in SYCL dialect. Are there a lot of deprecated features in sycl namespace?

**Comment by victor-eds** [comment]
> With this change, we may unexpectedly allow types in sycl namespace not in SYCL dialect. Are there a lot of deprecated features in sycl namespace?

Regarding types, only `sycl::atomic` AFAIK, but we still won't have support for all types in a long while and we may not even want that, as there are higher priority things than adding, e.g., `sycl::stream`

**Comment by whitneywhtsang** [comment]
> > With this change, we may unexpectedly allow types in sycl namespace not in SYCL dialect. Are there a lot of deprecated features in sycl namespace?
> 
> Regarding types, only `sycl::atomic` AFAIK, but we still won't have support for all types in a long while and we may not even want that, as there are higher priority things than adding, e.g., `sycl::stream`

As supporting the types are the ideal situation, I prefer to allow `sycl::atomic` explicitly in the assert like `h_item`. Change the default of `AllowUndefinedSYCLTypes` to true, and add a TODO comment above the option. WDYT?

**Comment by victor-eds** [comment]
Sorry, `sycl::atomic` is. There are other types (`atomic_ref`, `marray`, `reducer`, ...) which are not.

My concern here is **breaking** compilation just because we are compiling a class to `llvm.struct` instead of `sycl.foo`. Semantics should remain equivalent through both paths, but we'd just be missing SYCL-specific optimizations in the first one, ofc. Again, most of these cases are types we do not plan to support anytime soon.

I propose:
- Making default a compile-time option so we can easily switch and have local builds with that;
- CI still uses `true` default
- Add types to list

**Comment by victor-eds** [comment]
> CI still uses `true` default (not so sure about this, as we wouldn't be sure if tests are failing because one of those types or we're actually semantically incorrect)

I'd rather have `false` in CI by default, as that would give us more info on what's actually semantically incorrect in our code

**Comment by whitneywhtsang** [comment]
Sure, and let's add the known deprecated types to a "white" list, so no matter the option is true or false, the assertion won't be trigger by types we know we are not going to support.

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 1]
```diff
@@ -1281,6 +1284,27 @@ mlir::Type CodeGenTypes::getMLIRTypeForMem(clang::QualType QT,
...
 
+static bool isAllowedUndefinedSYCLType(StringRef TypeName) {
+  // Sorted list of allowed names
+  constexpr std::array<llvm::StringLiteral, 8> AllowedUndefinedSYCLType{
+      // Exceptions for `h_item` and `private_memory`: SYCL-MLIR doesn't
+      // support hierarchical parallelism as a first-class concept, but
+      // it can pass through its use to the runtime.
+      //
+      // For others: These should be semantically correct when lowered to
+      // llvm.struct. Use that untill the dialect is expanded.
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:1295]:
```suggestion
      // llvm.struct. Use that until the dialect is expanded.
```

**Review by whitneywhtsang — APPROVED** [review]

**Review by Naghasan — COMMENTED** [review]
LGTM, could you add a hook in the configure.py script

**Review by jopperm — APPROVED** [review]

**Pull request merged**